[체스 - 3, 4 단계] 짱수(장혁수) 미션 제출합니다.

zangsu opened the pull request:
안녕하세요 영이! 3, 4단계 미션 리뷰로 다시 찾아왔습니다!

우선 추가 요구사항들은 아직은 반영 전입니다.

추가 요구사항 : 
- 체스 게임방을 만들고 체스 게임방에 입장할 수 있는 기능을 추가한다.
- 사용자별로 체스 게임 기록을 관리할 수 있다.

### 핵심 변경사항
요구사항과 크게 관련되지 않은 변경사항입니다.

#### `Black/White Pawn` -> `Initial/Moved Pawn` 클래스로 변경

DB에 값을 저장하는 방법을 고민해 보던 중 기본 방법으로는 폰에 한정적으로 `boolean moved`를 저장해 주어야 했습니다. 이 때문에 데이터베이스에 저장하거나, 다시 기물을 생성하는 등의 상황에서 비효율적이라 생각했습니다.

이에 움직임의 상태를 클래스로 분리하였습니다. 

아래는 이와 관련된 추가 변경사항입니다. 
- `Piece.move()`가 `void`형 대신 `Piece`를 반환하게 되었습니다.
- `Color`가 각 진영의 진행 방향을 판단하는 책임을 가지게 되었습니다.

#### `Board`의 생성 책임을 분리

기존의 `Board`에서는 초기 생성을 위해 많은 코드가 존재했습니다. 
이를 별도의 팩토리 클래스로 분리하였습니다.

(이 때, 팩토리 클래스의 접근 제어자에 대해서 많이 고민을 했었는데요.
`default` 접근제어자를 사용하려고 하였으나, 과연 `default` 접근제어자를 사용하는 것 만으로 '같은 패키지의 `Board`를 위한 클래스, 메서드이다.' 라는 것을 나타낼 수 있을까 고민하다가 `public`으로 두었습니다.
`default` 접근 제어자가 의도를 표현하는 장점을 가지기보단, 소통 비용을 증가시키는 것 같다는 생각이 들기도 했어요.)

#### TODO 캐싱 삭제

캐싱을 고민을 했었는데요. DB와 OOP를 완벽하게 쳐내지 못하는 지금 상황에서 굳이 더 많은 것을 고민하는 것은 독이라고 생각했습니다. 😅 
또한, `Location` 객체가 64개의 한정적인 객체만을 사용하긴 하지만, 지금 굳이 캐싱을 고려해야 할 정도로 객체 생성 비용이 비싸다고 생각하지도 않고요.

만약 레벨1이 끝나기 전에 추가 요구사항까지 잘 끝낼 수 있다면 더 고민을 해 볼수도 있겠네요..ㅎ


### 마치며...

DB를 사용하는 것이 어색하다 보니 DB를 도입하는 시점에서 코드가 많이 지저분해진 것 같습니다.
이번에도 많은 피드백 부탁드립니다 🙇‍♂️

zangsu committed:
refactor (Board) : 예외를 조건문 내부에서 발생시키도록 변경

zangsu committed:
refactor (GameController) : ChessGame 객체를 필드가 아닌 지역변수로 가지도록 변경

zangsu committed:
refactor  : 가독성을 향상시키기 위한 메서드명 변경

zangsu committed:
refactor (Path) : 의미있는 매직넘버를 상수로 추출

zangsu committed:
refactor (Board) : 파라미터 값을 변경시키지 않도록 Location.copy() 메서드 사용

zangsu committed:
refactor (OutputView) : 각 기물의 출력 심볼을 Map 으로 관리하도록 변경

zangsu committed:
refactor (Row) : 접근제어자 변경

zangsu committed:
docs : 추가된 요구사항에 따라 README 작성

zangsu committed:
feat : 게임에 턴 개념을 도입

zangsu committed:
fix (GameController) : 잘못된 입력 이후 게임이 초기 상태로 돌아가던 버그 수정

zangsu committed:
feat (PlayingGame) : 왕이 죽으면 게임이 끝나도록 기능 수정

zangsu committed:
feat (Score) : 점수를 계산하는 기능을 구현

zangsu committed:
feat (Piece) : 각 기물 별 점수 구현

zangsu committed:
style : 도메인 용어를 반영하기 위해 Row -> Rank, Column -> File 로 변경

zangsu committed:
feat (Board) : 각 색깔 기물의 총 점수를 계산하는 기능 구현

zangsu committed:
feat (GameController) : 점수를 조회하는 기능 구현

zangsu committed:
style : 컨벤션 적용

zangsu committed:
style : 가독성을 위해 메서드 위치 정렬

zangsu committed:
feat : 초기 보드의 생성 책임을 클래스로 분리

zangsu committed:
feat : 데이터베이스 접근 기능 구현

zangsu committed:
feat (GameController) : 게임 종료 후 승자를 출력하는 기능 추가

- ChessGame.isEnd() 메서드의 필요에 따라 isNotEnd() -> isEnd() 로 변경
- 빈 Board 를 생성하는 정팩메 구현

zangsu committed:
feat(GameDao) : 게임을 저장하고 로드할 수 있는 기능 구현

zangsu committed:
feat(GameController) : 게임 저장 / 로드 기능 병합

zangsu committed:
refactor(GameController) : 게임이 종료된 이후에는 저장된 데이터를 삭제하도록 변경

zangsu committed:
refactor : 컨벤션 적용

zangsu committed:
style : TODO 삭제

- 지금 단계가 캐싱을 고민해야 하는 단계는 아니라고 생각
- Location 객체의 생성 비용이 캐싱을 고려해야 할 정도는 아니라고 생각

zangsu committed:
style : 해결된 TODO 삭제

zangsu committed:
refactor : 가독성을 위해 인라인 리턴문 사용

zangsu committed:
refactor(Piece) : Piece 가 움직인 이후 Piece 타입을 반환하도록 변경

zangsu committed:
refactor(Path) : 가독성을 위해 메서드명 변경

- 부정 확인 메서드명을 긍정 확인으로 변경
- Path 의 메서드명에 불필요한 Path 제거

zangsu committed:
refactor(Pawn) : Pawn 의 움직임 상태를 클래스로 구분하도록 변경

zangsu committed:
fix(create_table.sql) : 타입 이름의 추가로 인한 테이블 변경

zangsu committed:
fix(Pawn) : 진행 방향 이동시 상대편 말을 잡을 수 있던 오류 수정

zangsu committed:
refactor(GameController) : 예기치 못한 예외를 확인하기 위해 루트 예외 메시지 출력

zangsu committed:
refactor(InputView) : 잘못된 입력에 대한 재입력 요청 책임을 view에 위임

zangsu committed:
refactor : 가독성을 위한 메서드 정렬 및 이름 변경

zangsu committed:
style : 컨벤션 적용

zangsu commented on src/main/java/chess/repository/GameDao.java:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    private void initializeBoardTable(Connection connection) throws SQLException {
+        //TODO 여러 게임을 관리하게 되면 꼭 WHERE 절이 필요해진다.
```
추가 요구사항 구현 (레벨 5라 칭하겠습니다.) 에서 꼭 신경을 써야 할 부분일 것 같아 TODO를 작성한 채 리뷰를 올리게 되었습니다.

zangsu committed:
refactor(GameDao) : 쿼리를 한 번에 전송하도록 변경

zangsu committed:
refactor(GameDao) : 인덴트 조건을 위해 메서드 추출

zangsu commented on src/main/java/chess/domain/board/Step.java:
```diff
@@ -33,11 +29,19 @@ public boolean hasEnemy() {
     }
 
+    public boolean hasAlly() {
+        return locationState.isAlly();
```
이런 부분들은 사실 다음과 같이 작성해도 될 것 같은데요.
```suggestion
        return locationState == LocationState.ALLY;
```

최대한 객체에 메시지를 보내도록 구현하기 위해 조금 과하게 메서드를 만든 것 같기도 합니다.
두 개의 차이가 있을까요?? 이런 상황에도 객체에 메시지를 보내도록 해야 할까요?

choijy1705 replied:
== 비교로도 충분할 것 같아요!
그냥 취향 차이 정도 일것 같네요

zangsu reviewed (commented):
고민 및 참고사항들도 공유드립니다!

choijy1705 commented on docker/db/mysql/init/create_table.sql:
```diff
@@ -0,0 +1,10 @@
+CREATE TABLE BOARD
```
```suggestion
CREATE TABLE BOARDS
```
테이블 명은 보통 복수로 관리할 것 같아요!

choijy1705 replied:
테이블마다 id는 필요없을까요?

zangsu replied:
지금 당장에는 하나의 게임만을 저장하고, 때문에 하나의 보드 정보만을 저장하게 될 것 같아서 `BOARD`라는 네이밍을 했습니다. 동일하게 하나의 게임만을 저장하기 때문에 `id` 값 역시 다음 단계에서 추가를 하는 것이 맞을 것 같다고 생각을 했고요.

그러나, 이렇게 구현을 하니 특히 `GAME` 테이블이 1개의 컬럼과 1개의 로우만을 가지게 되더라고요... 😅 
조금은 5단계 구현을 염두에 두는 방법이겠지만, 복수형을 사용하고 `id`도 추가하는 방향으로 변경해 볼게요!
(어떻게 보면 이정도의 확장성 고려는 오히려 자연스러워 보이기도 하네요)

choijy1705 commented on docker/db/mysql/init/create_table.sql:
```diff
@@ -0,0 +1,10 @@
+CREATE TABLE BOARD
+(
+    Location  char(2) NOT NULL ,
```
```suggestion
    location  char(2) NOT NULL ,
```

choijy1705 commented on src/main/java/chess/repository/GameDao.java:
```diff
@@ -0,0 +1,153 @@
+    private static final String DATABASE = "chess"; // MySQL DATABASE 이름
+    private static final String OPTION = "?useSSL=false&allowPublicKeyRetrieval=true&serverTimezone=UTC";
+    private static final String USERNAME = "root"; //  MySQL 서버 아이디
+    private static final String PASSWORD = "root"; // MySQL 서버 비밀번호
```
ym 혹은 properties 파일에서 관리되면 좋을것 같네요

choijy1705 commented on src/main/java/chess/controller/GameController.java:
```diff
@@ -11,63 +12,94 @@
+    @FunctionalInterface
+    private interface GameStateChanger {
+        ChessGame change(ChessGame game);
+    }
```
가독성이 좋아졌다고 느끼셨을까요??
이전 방식대로 contrller를 생성할때 ChessGame을 초기화해주고 이후 로직을 처리하는것이 일반적일것 같습니다 ㅎㅎ

zangsu replied:
함수형 인터페이스와 `Map<Command, GameStateChanger>`를 사용하는 대신 `switch` 문에서 어떤 메서드가 실행되는지 명시하도록 변경하였습니다.

choijy1705 commented on src/main/java/chess/controller/GameController.java:
```diff
@@ -11,63 +12,94 @@
 public class GameController {
     private static final InputView INPUT_VIEW = new InputView();
     private static final OutputView OUTPUT_VIEW = new OutputView();
+    private static final GameDao GAME_DAO = new GameDao();
```
dao는 왜 상수로 관리하나요??

zangsu replied:
아하...! Dao 를 상수로 관리하는 순간 상위 클래스가 하위 클래스의 특정 구현체와 강결합이 되겠네요!
지금의 경우는 Dao도 구현체이지만, 실제로는 Dao가 하나의 인터페이스로 정의되고 사용 데이터베이스마다 구현체가 있다고 들었던 것 같아요. 

필요에 따라 구현체를 갈아끼우기 위해선 외부에서 의존성을 주입받아야 할 것 같네요.
리팩토링 진행하겠습니다!

choijy1705 commented on src/main/java/chess/controller/GameController.java:
```diff
@@ -11,63 +12,94 @@
         } catch (IllegalArgumentException | IllegalStateException exception) {
             OUTPUT_VIEW.printExceptionMessage(exception.getMessage());
-            runOrRetry();
+            return playTurn(chessGame);
         } catch (RuntimeException exception) {
-            OUTPUT_VIEW.printExceptionMessage("예기치 못한 동작입니다. 다시 명령어를 입력해 주세요.");
-            runOrRetry();
+            OUTPUT_VIEW.printExceptionMessage("예기치 못한 동작입니다. " + exception.getMessage());
+            return playTurn(chessGame);
         }
```
RuntimeException 으로만 catch해도 되지 않을까요?

zangsu replied:
이 부분이 두 가지 `catch` 문으로 나뉘었던 이유는 이전의 영이 피드백을 듣고 의도적으로 발생시키지 않은 예외에 대해 조금 더 사용자 친화적인 메시지를 출력해 주기 위함이었는데요.
[[1, 2단계 - 체스] 짱수(장혁수) 미션 제출합니다.](https://github.com/woowacourse/java-chess/pull/640#discussion_r1536623349)

다시 생각해 보니 `RuntimeException`을 잡는 상황이라면 유저의 잘못된 입력이 문제가 되는 경우가 아닐 것 같아 재귀함수를 호출할 이유는 없는 것 같네요. 영이도 이 때문에 다시 리뷰를 주신 것 같아서 이곳에서 처리해 주는 대신, 조금 더 상위 메서드로 이동 후 프로그램을 종료시키는 방향으로 리팩토링을 해 보겠습니다.

choijy1705 commented on src/main/java/chess/repository/PieceDBMapper.java:
```diff
@@ -0,0 +1,48 @@
+import chess.domain.piece.implement.pawn.MovedPawn;
+import java.util.Arrays;
+
+public class PieceDBMapper {
```
해당 클래스의 역할은 뭔가요?
네이밍이 적절한지 고민해보면 좋을것 같습니다

zangsu replied:
범용적으로 사용되는 `XXXFactory` 네이밍을 사용하도록 변경해 보았습니다.

choijy1705 commented on src/main/java/chess/repository/GameDao.java:
```diff
@@ -0,0 +1,153 @@
+import java.util.Map;
+import java.util.Optional;
+
+public class GameDao {
```
BoardDao는 필요없나요?

choijy1705 commented on src/main/java/chess/repository/GameDao.java:
```diff
@@ -0,0 +1,153 @@
+    private static final String USERNAME = "root"; //  MySQL 서버 아이디
+    private static final String PASSWORD = "root"; // MySQL 서버 비밀번호
+
+    public Connection getConnection() {
+        // 드라이버 연결
+        try {
+            return DriverManager.getConnection("jdbc:mysql://" + SERVER + "/" + DATABASE + OPTION, USERNAME, PASSWORD);
+        } catch (final SQLException e) {
+            System.err.println("DB 연결 오류:" + e.getMessage());
+            e.printStackTrace();
+            return null;
+        }
+    }
```
이 부분은 여기서 있어도 괜찮을까요??
dao는 일반적으로 여러개일것 같아서요!

choijy1705 commented on src/main/java/chess/repository/GameDao.java:
```diff
@@ -0,0 +1,153 @@
+        }
+    }
+
+    public void saveGame(ChessGame playingGame) {
+        try (Connection connection = getConnection()) {
+            initializeGameTable(connection);
+            initializeBoardTable(connection);
+            saveGameData(playingGame, connection);
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private void saveGameData(ChessGame playingGame, Connection connection) throws SQLException {
+        initializeGameTable(connection);
+        initializeBoardTable(connection);
+        if (playingGame.isEnd()) {
+            return;
+        }
+        saveTurn(connection, playingGame.getTurn());
+        saveBoardToDB(connection, playingGame.getBoard());
+    }
```
로직이 중복되는 부분이 많을것 같은데 고민해보면 좋을것 같아요!

choijy1705 commented on src/main/java/chess/repository/GameDao.java:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    private void saveBoardToDB(Connection connection, Map<Location, Piece> board) throws SQLException {
+        PreparedStatement preparedStatement = connection.prepareStatement("insert into board values (?, ?, ?)");
```
여기는 쿼리가 소문자에요
일관성있게 가져가보면 좋을것 같습니다!

choijy1705 commented on src/main/java/chess/repository/GameDao.java:
```diff
@@ -0,0 +1,153 @@
+
+    private void initializeBoardTable(Connection connection) throws SQLException {
+        //TODO 여러 게임을 관리하게 되면 꼭 WHERE 절이 필요해진다.
+        PreparedStatement preparedStatement = connection.prepareStatement("DELETE FROM BOARD");
```
gameDao인데 board로 쿼리를 날리고 있네요!

choijy1705 replied:
dao가 지저분해지는 큰 이유중하나가 모든 db 관련 로직을 gameDao에서 처리하고 있습니다.(심지어 connection 까지도....)
gameDao와 boardDao를 적절히 나눠보고 두 dao를 통해 조회한 값을 통해 비즈니스 로직을 처리할 
service 레이어까지 고려해보면 좋을것 같아요!

choijy1705 commented on src/main/java/chess/domain/board/InitialBoardBuilder.java:
```diff
@@ -0,0 +1,67 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class InitialBoardBuilder {
```
BoardFactory로 명명하는게 어떨까요?

zangsu replied:
빌더와 팩토리 모두 널리 쓰이는 디자인 패턴이니 팩토리로 변경해 주는 것이 훨씬 직관적이겠네요!

choijy1705 commented on src/main/java/chess/repository/GameDao.java:
```diff
@@ -0,0 +1,153 @@
+        preparedStatement.executeBatch();
+    }
+
+    public Optional<ChessGame> loadGame() {
```
dao에서 CRUD 사용시 일반적인 네이밍이 있는것 같아요
Create -> save or create
Read -> find
Update -> update
Delete -> delete

choijy1705 commented on src/main/java/chess/repository/GameDao.java:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    private void loadBoard(Connection connection, Map<Location, Piece> board) throws SQLException {
+        PreparedStatement preparedStatement = connection.prepareStatement("select * from board");
```
여기도 board를 gameDao에서 조회하네요....

choijy1705 reviewed (changes requested):
짱수 안녕하세요!
GameDao에서 대부분의 로직들을 처리하고 있는것 같아요!
테이블은 board와 game 두가지인데 game에서 board 를 처리하고 있는건 많이 어색한것 같네요!
이부분 중점적으로 수정해보면 좋을것 같아요!

zangsu committed:
refactor(InitialBoardFactory) : 가독성을 위해 XXXBuilder -> XXXFactory 로 클래스명 변경

zangsu committed:
style : 테이블 이름 및 컬럼 명 변경, SQL 쿼리문 대소문자 통일 및 이름 변경

zangsu committed:
refactor (DatabaseUtil) : DB Connection 얻어오는 기능 수정

- 데이터베이스 설정 값을 yml 파일로 관리하도록 변경
- DAO 분리를 고려해 Connection 을 얻어오는 역할 분리

zangsu committed:
refactor (Controller) : 의도하지 않은 예외에 대해 예외 처리를 하는 대신 예외 메시지만 출력 후 프로그램 종료

zangsu committed:
refactor (ChessGame) : getBoard() 메서드가 도메인 객체 Board 를 반환하도록 변경

zangsu committed:
refactor (GameService) : DAO 를 분리하며 DB 에서 조회한 값을 다루기 위한 Service 계층 구현

- Controller 가 가지고 있던 DAO 객체를 Service 로 이동
- 각 테이블마다 DAO 를 가지도록 분리
- Service, Dao 객체의 의존성을 외부에서 주입

zangsu committed:
refactor : domain 패키지를 Service 패키지 하위로 이동

zangsu committed:
refactor(GameController) : 가독성을 위해 각 커멘드마다 수행될 메서드를 직접 switch-case 에서 명시

zangsu commented on src/main/java/chess/controller/GameController.java:
```diff
@@ -1,73 +1,101 @@
-                    .run();
-        }
+    private ChessGame executeCommand(ChessGame chessGame, Command command) {
+        return switch (command) {
```
>- else 예약어를 쓰지 않는다.
>   - else 예약어를 쓰지 말라고 하니 switch/case로 구현하는 경우가 있는데 switch/case도 허용하지 않는다.

아시겠지만, 우테코에서는 `switch-case` 문을 허용하지 않습니다. 

`else`를 사용한다면 `else`에 해당하는 조건을 확인하기 위해 위의 모든 조건을 확인할 필요가 있기 때문에 `else`를 허용하지 않는 것을 이해했는데요.
그렇다면, `switch`문에서 역시 `default`를 사용하지 않고 모든 케이스를 명시해 준다면 `switch-case`를 사용하지 않을 이유가 없다고 생각합니다. 

또한 `switch`문에서 enum을 사용한다면 모든 케이스를 명시해 주지 않을 경우 컴파일 에러가 나기에 모든 enum값에 대해 다른 결과를 사용해야 하는 경우 `switch`문을 사용하게 되었습니다.

zangsu commented on src/main/java/chess/repository/YmlFile.java:
```diff
@@ -0,0 +1,63 @@
+import java.util.regex.Pattern;
+
+
+public class YmlFile {
```
Yml 파일은 상황에 따라 충분히 depth 가 깊어질 수 있을텐데요.
우선, 지금 상황에서 다중 depth 의 yml 파일을 관리할 일은 없을 것이라 생각해 single depth인 Yaml 파일을 파싱하기 위한 클래스를 만들어 보았습니다.

물론, 어떻게 보면 그냥 파싱만 하면 될 일을 `YmlFile` 까지 만드는 것이 과하게 보이기도 하는데요...ㅎㅎ
'왜 이렇게 과하게 구현했나' 라고 물어보신다면 욕심이 나서 정신차려 보니 여기까지 오게 되었다 라는 말 밖에 못드릴 것 같네요...ㅠ

zangsu committed:
refactor(PropertiesFile) : 설정 파일의 1 depth 고정을 위해 yaml 대신 properties 파일 사용

zangsu committed:
refactor(PropertiesFile) : 설정 파일의 1 depth 고정을 위해 yaml 대신 properties 파일 사용

zangsu commented on src/main/java/chess/repository/PropertiesFile.java:
```diff
@@ -0,0 +1,65 @@
+import java.util.regex.Pattern;
+
+
+public class PropertiesFile {
```
어떻게 보면 그냥 파싱만 하면 될 일을 특정 파일 클래스까지 만드는 것이 과하게 보이기도 하는데요...ㅎㅎ
'왜 이렇게 과하게 구현했나' 라고 물어보신다면 욕심이 나서 정신차려 보니 여기까지 오게 되었다 라는 말 밖에 못드릴 것 같네요...ㅠ

zangsu committed:
feat(GameDao, BoardDao) : 게임 데이터에 게임 식별을 위한 아이디를 추가

zangsu committed:
feat(ChessGame) : 게임 객체가 게임 식별을 위한 아이디를 가지도록 변경

zangsu committed:
fix(PropertiesFile) : properties 파일의 value 에 '='을 포함할 수 있도록 변경

zangsu committed:
refactor(ChessGame) : 게임 Id를 가지도록 도메인 변경

zangsu committed:
fix(database.properties) : 글자 깨짐 문제를 해결하기 위해 주석을 영어로 변경

zangsu committed:
refactor(PieceGenerator) : 책임을 잘 나타낼 수 있는 이름으로 클래스명 변경

zangsu committed:
style : 컨벤션 적용

zangsu committed:
chore : 깃 충돌 해결

zangsu committed:
refactor : 디버깅에 사용했던 출력문 제거

zangsu committed:
feat(GameController) : 게임 종료 기능과 저장 기능 분리

zangsu committed:
feat(OutputView) : 게임 저장 기능에 대한 출력 추가

zangsu committed:
style : 가독성을 위해 인라인 리턴

zangsu commented on src/main/java/chess/service/domain/chessGame/ChessGame.java:
```diff
@@ -0,0 +1,36 @@
+import chess.service.domain.piece.Score;
+import java.util.function.Supplier;
+
+public abstract class ChessGame {
```
### 변경 내역

DB에서 게임 Id를 저장하게 되면서 도메인에도 Id 값을 가지고 있도록 변경해야 했습니다.
추상클래스가 도입이 될 경우 기존의 인터페이스의 존재 이유가 흐려지는 것 같아 기존의 `ChessGame` 인터페이스를 추상클래스로 변경하였습니다.

### 변경 예정

크루들과 이야기를 나누던 도중 저의 `ChessGame` 인터페이스 (지금은 추상클래스)가 리스코프 치환 원칙을 지키지 못한다는 점을 발견했습니다. 물론 SOLID 원칙은 꼭 지켜야 하는 것은 아니지만, 학습하는 지금은 엄격하게 지켜 보는 것도 좋을 것 같다고 생각했어요. 

다음 리뷰 반영때 `ChessGame`에 적용되어 있던 상태 패턴을 제거할 예정입니다.
(코드 변경이 너무 많아지면 제가 리뷰를 요청드리는 주기도 길어지고, 영이도 코드 변경을 파악하는 데 더 부담이 들 것으로 생각해 다음 리뷰로 미루게 되었습니다. 리뷰가 도착할 때 까지 열심히 구상해 보겠습니다!)

zangsu replied:
### 리스코프 치환 원칙

관련해서 오늘 구구랑 이야기를 해 보았는데요.
구구는 상위 메서드에서 굳이 특정 예외 상황을 명시해 두지 않아도 (즉 이전 상태의 `ChessGame`도) 충분히 리스코프 치환 원칙을 지키고 있다고 생각한다 말해 주셨어요.
상위 메서드의 의도대로 동작을 하지 못하는 상황에서 예외를 던져 주는 것이 왜 문제인가? 라는 의견을 주셨어요.

비슷한 결로 `Collections.UnmodifiableList`는 리스코프 치환 원칙을 지키고 있나? 라는 논쟁이 SOF에 있더라고요. 혹시나 영이가 궁금해 하실까 싶어 함께 첨부해 둡니다. 
[Do Collections.unmodifiableXXX methods violate LSP?](https://stackoverflow.com/questions/22050848/do-collections-unmodifiablexxx-methods-violate-lsp)

리스코프 치환 원칙에 대해 다시 조금 헷갈리기 시작하는 것 같아요. 
지금 당장의 저는 위의 SOF 가장 위에 있는 답변에 가장 동의합니다. 상위 메서드에서 명시해 둔 예외를 사용하는 상황에 한정적으로 리스코프 치환 원칙을 준수하고 있다고 생각해요. 

### throws

해당 이야기를 나누면서 언체크 예외를 `throws` 로 명시해 주는 것에 대한 피드백을 받았어요. 그래서 `throws`를 지워 주는 대신 Javadoc을 사용해 예외를 명시해 주는 방법으로 변경해 보았습니다. 

물론, 주석을 사용하는 방법이기에 매 변경마다 싱크를 맞춰 주어야 한다는 단점이 있는데요. 
그럼에도 불구하고 언체크 예외를 `throws`로 명시하여 주는 혼란에 비해 인지적 부담이 현저히 적으며, 인터페이스에서는 해당 메서드의 의도를 잘 설명해 주는 것이 좋을 것 같다는 생각에 Javadoc을 사용했습니다.

zangsu committed:
style : 가독성을 위해 클래스 이름 변경 및 메서드 정렬,정리

choijy1705 commented on docker/db/mysql/init/create_table.sql:
```diff
@@ -0,0 +1,14 @@
+CREATE TABLE GAMES(
+    game_id int NOT NULL,
```
```suggestion
    id int NOT NULL,
```
이미 게임이라는 테이블이기 때문에 id라고 명명하는게 적절할것 같습니다!

choijy1705 replied:
id의 경우 BIGINT를 사용하는게 어떨까요?

zangsu replied:
식별자는 조금 널널하게 값의 범위를 제공해 주는 것이 좋겠네요!

choijy1705 commented on docker/db/mysql/init/create_table.sql:
```diff
@@ -0,0 +1,14 @@
+
+CREATE TABLE BOARDS
+(
+    game_id int NOT NULL,
```
```suggestion
    id int NOT NULL,
```
여기도 마찬가지로요!

zangsu replied:
BOARD 테이블에서는 외래키로 가지는 id 값이 어디의 id 인지 알 수 있어야 하지 않을까요?
단순히 id 라고만 적어주면 BOARD의 id 값인지 GAME의 id 값인지 헷갈릴 것 같아요.

choijy1705 replied:
외래키로 쓰이는 game_id군요 그렇다면 더더욱 이상한것 같습니다
piece 테이블에는 id가 필요없을까요?

choijy1705 commented on src/main/java/chess/repository/PropertiesFile.java:
```diff
@@ -0,0 +1,69 @@
+            throw new IllegalArgumentException("properties 파일이 아닙니다.");
+        }
+        try {
+            BufferedReader bufferedReader = new BufferedReader(new FileReader("./src/\\main\\resources\\" + fileName));
```
```suggestion
            BufferedReader bufferedReader = new BufferedReader(new FileReader("./src/main/resources/" + fileName));
```
실행이 안되고 있는것 같네요!

zangsu replied:
수정했습니다!

choijy1705 commented on src/main/java/chess/repository/DatabaseConnectionGenerator.java:
```diff
@@ -0,0 +1,42 @@
+        } catch (final SQLException e) {
+            System.err.println("DB 연결 오류:" + e.getMessage());
+            e.printStackTrace();
+            return null;
```
왜 throw 하지 않고 null을 반환하나요??

zangsu replied:
예시 코드를 사용 후 나중에 바꿔야지, 하고 놓쳤네요 😞 
수정했습니다!

choijy1705 commented on src/main/java/chess/repository/BoardDao.java:
```diff
@@ -0,0 +1,90 @@
+        PreparedStatement preparedStatement = connection.prepareStatement(
+                "INSERT INTO BOARDS (game_id, location, piece_type, color) VALUES (?, ?, ?, ?)"
+        );
+        for (Location location : board.keySet()) {
+            Piece piece = board.get(location);
+
+            String locationData = location.getFile().getSymbol() + location.getRank().getSymbol();
+            preparedStatement.setInt(1, gameId);
+            preparedStatement.setString(2, locationData);
+            preparedStatement.setString(3, piece.getPieceType().name());
+            preparedStatement.setString(4, piece.getColor().name());
+
+            preparedStatement.addBatch();
+            preparedStatement.clearParameters();
+        }
```
로직적으로 문제가 되는건 아니지만 entrySet을 사용하지 않고 
keySet으로 for문을 만들고 다시 map에서 get하는 방법은 조금 어색한 방법인것 같네요!

zangsu replied:
수정했습니다!

choijy1705 commented on src/main/java/chess/repository/GameDao.java:
```diff
@@ -0,0 +1,92 @@
+        return new Game(gameId, turn);
+    }
+
+    public void updateGameById(Game game) {
+        try (Connection connection = connectionGenerator.getConnection()) {
+            PreparedStatement preparedStatement = connection.prepareStatement(
+                    "UPDATE GAMES SET turn = ? WHERE game_id = ?");
+            preparedStatement.setString(1, game.getTurn().name());
+            preparedStatement.setInt(2, game.getGameId());
+            preparedStatement.execute();
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
+    }
```
안쓰이네요!

choijy1705 commented on src/main/java/chess/service/GameService.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    public ChessGame createNewGame() {
+        int lastGameId = gameDao.findLastGameId();
+        //TODO 방을 여러개 관리하게 되면 이전 데이터를 남겨두고 새로 만들어야 함
+        boardDao.deleteBoardById(lastGameId);
+        gameDao.deleteGameById(lastGameId);
+        if (lastGameId == 0) {
+            lastGameId = 1;
+        }
+        return new InitialGame(lastGameId);
```
createNewGagme 인데
game을 만들고 db에 저장하는 로직은 없는것 같아요!

choijy1705 commented on src/main/java/chess/service/GameService.java:
```diff
@@ -0,0 +1,69 @@
+        return chessGame;
+    }
+
+    public void save(ChessGame chessGame) {
+        Game game = new Game(chessGame.getGameId(), chessGame.getTurn());
+        boardDao.deleteBoardById(game.getGameId());
+        gameDao.deleteGameById(game.getGameId());
+        gameDao.saveGame(game);
+        boardDao.saveBoard(game.getGameId(), chessGame.getBoard());
+    }
```
게임이 진행될때마다 저장되는것이 아닌
save를 눌러야 게임이 저장되는군요...?
조금 옛날 방식의 게임 진행방식일것 같네요! 변화가 있을때마다 저장되도록 할 수 는 없을까요?

zangsu replied:
영이 말 대로 움직임마다, 또는 게임 시작마다 DB 에 자동으로 저장할 수 있도록 변경해 보았어요. 

+) 어제 DM으로 `Board`를 지워도 되겠냐고 물어봤던 것은 조금 생각이 꼬였던 것 같아요.
다시 생각을 정리해 본 결과 제가 원했던 것은 `Controller`에 지역 변수로 존재하고 있는 `ChessGame`을 없애고, 게임 Id만을 `Controller`가 가지도록 변경하고 싶었던 것 같아요. 

매 순간 게임의 정보가 DB에 저장되기 때문에 `Controller`에서 `ChessGame`을 지역 변수로 관리하지 않더라도 DB에서 Id 값으로 조회를 할 수가 있으니까요.
이 방법에 장점은 커맨드 입력들을 비연결적으로 제공할 수 있는 점이에요. 
(물론 지금 당장에는 `Controller`가 `int gameId`를 가지고 있어야 하기 때문에 완전한 비연결성은 아니지만, 추후 게임 아이디를 외부에서 받아올 수 있게 된다면 손쉽게 비연결성을 제공할 수 있을 것 같아요.)

하지만, 결과적으로 그렇게 리팩토링 하지 않은 이유는 지금은 콘솔 프로그램이기 때문이에요. 
많은 사용자가 동시에 접속하지 않는 현재 상황에서 굳이 비연결성을 제공해 줄 필요도 없거니와, 그렇게 변경하면 매 움직임마다 DB에 접근해서 게임을 복원해야 하는 것이 효율적이지 못하게 느껴졌어요.

현재의 객체 상태를 최대한 유지하는 방향으로 진행했습니다.

choijy1705 commented on src/main/java/chess/service/GameService.java:
```diff
@@ -0,0 +1,69 @@
+
+    public void save(ChessGame chessGame) {
+        Game game = new Game(chessGame.getGameId(), chessGame.getTurn());
+        boardDao.deleteBoardById(game.getGameId());
+        gameDao.deleteGameById(game.getGameId());
+        gameDao.saveGame(game);
+        boardDao.saveBoard(game.getGameId(), chessGame.getBoard());
```
트랜잭션에 관한 고민도 해보면 좋을것 같습니다!

zangsu replied:
서비스 계층에서 트랜잭션을 관리하도록 변경해 보았습니다.
필요한 모든 메서드에서 `try-catch-finally`가 반복될 것 같아 해당 부분을 추출하여 `TransactionManager`를 만들었습니다.

choijy1705 reviewed (changes requested):
짱수 많은 것들을 고민해보고 시도해보시고 계신것 같네요👍
추가 코멘트 드렸습니다!
확인해보시고 반영부탁드립니다!

zangsu committed:
refactor(GameDao) : 테이블 변경

- 테이블 이름과 중복되는 컬럼병 변경
- SQL 명령어는 대문자로 통일
- id 값에 BIGINT 범위 제공

zangsu committed:
fix(PropertiesFile) : 설정 파일의 상대 경로 수정

zangsu committed:
fix(DatabaseConnectionGenerator) : DB 연결 실패시 null 을 반환하는 대신 예외 발생

zangsu committed:
refactor(BoardDao) : keySet -> entrySet 으로 변경

zangsu committed:
style(PieceDao) : 피스 정보들을 저장한다는 책임을 직관적으로 나타내기 위해 Dao, Table 의 이름에 Piece 사용

zangsu committed:
feat(GameService) : 매 움직임마다 DB에 상태를 저장하도록 변경

zangsu committed:
refactor(Command) : 자동 저장으로 필요없어진 Save 커맨드 제거

zangsu committed:
refactor(ChessGame) : 리스코프 치환 원칙을 준수하기 위해 상위 클래스에서 반환 예외를 명시

zangsu committed:
style : 컨벤션 적용

zangsu committed:
refactor(TransactionManager) : Service 에서 트랜잭션을 관리할 수 있도록 변경

zangsu committed:
test(EndGameTest) : 종료된 게임 상태에서도 종료, 시작을 할 수 있도록 테스트 변경

zangsu committed:
refactor : Dao 메서드 파라미터로 도메인을 사용하도록 변경

zangsu committed:
refactor : 자원 사용 후 반환하도록 변경

zangsu commented on src/main/java/chess/service/domain/chessGame/ChessGame.java:
```diff
@@ -0,0 +1,39 @@
+
+    public abstract ChessGame endGame();
+
+    public abstract ChessGame move(Location source, Location target) throws NotPlayingGameException;
```
우선 각 메서드가 제대로 동작할 수 없는 예외 상황을 명시해 두고, 하위 타입이 해당 예외를 반환할 수 있도록 하여 리스코프 치환 원칙을 준수하도록 변경해 보았습니다.
이에 상태 패턴을 제거하지 않는 방향으로 진행하고자 합니다.

+) `ChessGame` 클래스들을 조금 더 개선할 수 있을 것 같은데, 가닥이 잡히지 않는 것 같습니다. 
(대표적으로 `ChessGame`을 상속한 `initialGame`의 메서드들이 대부분 예외를 던지고 있어 이 부분에서 인지적 부담이 존재한다는 크루들의 의견이 있었는데요, 이를 해결하기 위해선 상태패턴을 제거하고 하나의 `ChessGame`에서 `if` 분기를 사용하는 방법밖에 없을까요?)

이 부분을 포함하여 전체적으로 영이가 보시기에 아쉬운 점이 있으시다면 한번 더 RC 부탁드립니다.
매번 감사합니다! 🙇‍♂️

zangsu commented on src/main/java/chess/repository/TransactionManager.java:
```diff
@@ -0,0 +1,42 @@
+        try {
+            execute(transaction);
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
```
`SQLException`이 발생할 경우 개발자가 예외 처리를 해줄 수 있는 방법이 없을 것이라 여겨 `RuntimeException`으로 감쌌습니다.
최종적으론 예외 메시지 출력 후 프로그램이 종료됩니다.
다만, 외부 메서드에 `throws SQLExceptions`를 명시하는 대신 `RuntimeException`으로 감싸는 것이 괜찮은 선택인지는 모르겠네요...

choijy1705 replied:
지금 방법도 괜찮은 방법이에요👍
RuntimeException 을 상속하고 SQL 관련 exception 이라는 것을 알 수 있을만한 새로운 exception을 만들어서 던지면 더 좋을것 같네요!

zangsu commented on src/main/java/chess/repository/GameDao.java:
```diff
@@ -0,0 +1,81 @@
+        this.connectionGenerator = connectionGenerator;
+    }
+
+    public void saveGame(Connection connection, ChessGame game) throws SQLException {
```
저장, 업데이트 등을 하는 시점에서 DAO가 도메인을 몰라야 할 이유를 찾이 못해 도메인 객체를 그대로 넘겨주게 되었습니다.
조회의 상황에서만 제한적으로 하나의 클래스가 하나의 DAO에서 만들어 지기 어려워 `Entity`를 사용하게 되었습니다.

choijy1705 replied:
저장하는 시점에 도메인을 아는게 당연하지 않나요?
도메인이 항상 우선시되고 도메인을 통해서 db에 저장되어야할것 같습니다!
물론 entity가 필요한 경우가 있습니다. 
도메인이 엄청 복잡해지고 여러 dao를 통해서 하나의 도메인이 만들어지는 경우 entity가 필요할것 같은데 
미션 단계에서는 그런 경우는 없을것 같습니다🤔

zangsu committed:
style : ChessGame 에 Javadoc 추가

choijy1705 commented on docker/db/mysql/init/create_table.sql:
```diff
@@ -0,0 +1,14 @@
+    PRIMARY KEY (id)
+);
+
+CREATE TABLE PIECES
+(
+    game_id BIGINT NOT NULL,
+    location  CHAR(2) NOT NULL ,
+    piece_type VARCHAR(12) NOT NULL,
+    color CHAR(5) NOT NULL,
+    FOREIGN KEY (game_id) REFERENCES GAMES(id)
+);
```
테이블을 만들때 언제나 테이블 값과 무관한 key의 역할을 하는 id가 필요하다고 생각합니다
한 예로 옛날에 주민등록번호를 pk로 하고 별도의 id를 따로 만들지 않았던 회사가 잇었는데
정부 정책으로 주민번호를 저장하지 못하게 되면서 테이블 전부를 갈아 엎어야했던 경우가 있을것 같네요

zangsu replied:
아! 바로 납득이 되었습니다! 
PK 추가하겠습니다!

choijy1705 commented on src/main/java/chess/service/domain/board/Board.java:
```diff
@@ -0,0 +1,148 @@
+package chess.service.domain.board;
```
왜 서비스 안에 도메인이 들어있나요?😭

zangsu replied:
앗, 덕분에 다시 공부해 볼 수 있게 되었네요!!
지금까지 Service 레이어의 역할이 도메인들과 소통하여 비즈니스 로직을 처리하는 계층이며, 때문에 domain은 Service 레이어에 속해있다고 생각했던 것 같아요.

그냥 도메인은 레이어드 아키텍쳐랑은 별개로 해당 프로그램에서 의미있는 책임을 가지는 객체들이 되겠네요. 
도메인과 서비스 패키지를 분리하겠습니다!

choijy1705 commented on src/main/java/chess/service/domain/piece/Piece.java:
```diff
@@ -0,0 +1,65 @@
+import chess.service.domain.board.Path;
+import java.util.Objects;
+
+public abstract class Piece {
```
DB의 컬럼과 도메인의 필드가 불일치 하는것 같아요
실무에서는 물론 일치하지 않는 경우가 많고 반드시 일치할 필요가 없는 부분인것은 맞지만
지금 미션단계에서 이렇게 차이가 나버리면 나중에 더 복잡한 도메인에서는 어떤게 어떤건지 모르는 상황이 발생할것 같습니다!

choijy1705 replied:
DM으로 전달 드렸었던것 같은데 한번 고민해보면 좋을 포인트 남겨놓을께요
1. Board 테이블은 필요 없을까요? 
2. DB에는 piece 테이블에 location 이 있는데 도메인은 Board에서 location을 관리하고 있습니다
3. 도메인이 우선일지 entity가 우선일지 생각해보면 좋을것 같아요. 도메인이없는 entity Game 이 필요할지?
혹은 entity 와 도메인에 대한 정의가 무엇일지 생각해보면 좋을것 같네요

zangsu replied:
1, 2 번 질문에 대한 답을 별도의 코멘트로 작성해 두었습니다! 답을 찾는 과정에서 고민이 길어져 조금 늦었네요 ㅠㅠ

3번 질문의 의도를 고민하던 중 `Game` 엔티티가 없어도 아무런 문제가 없다는 사실을 깨달았습니다!!
어쩌면 지금까지 '도메인을 DB에 저장한다' 가 아닌, 'DB 에 저장된 값을 사용한다.' 라는 관점으로 이번 단계를 진행하고 있었나봐요!!

때론 특정 책임을 가지는 도메인으로 구현하기엔 애매한 작은 단위의 정보들이 Entity로 존재할 수 있겠지만, 적어도 이번 미션의 경우 그렇지 않을 것 같아요.
아직 '모든 것을 객체로 생각하는 연습'이 더 필요할 것 같네요

choijy1705 commented on src/main/java/chess/service/GameService.java:
```diff
@@ -0,0 +1,89 @@
+    public ChessGame createNewGame() {
+        int lastGameId = gameDao.findLastGameId();
+        //TODO 방을 여러개 관리하게 되면 이전 데이터를 남겨두고 새로 만들어야 함
+        transactionManager.executeTransaction(
+                connection -> {
+                    pieceDao.deleteAllPiecesById(connection, lastGameId);
+                    gameDao.deleteGameById(connection, lastGameId);
+                });
```
좋은 시도였습니다👍
다만 가독성의 측면에서 좋을지 고민해보면 좋을것 같아요

zangsu replied:
가독성이 그렇게 좋지는 못하다는 의견에 동의합니다.
메서드 체이닝을 구현해 볼까 어쩔까 고민을 해보았지만, 구현을 성공하진 못했네요 😢
방학때 심심해 진다면 조금 더 고민해 볼지도 모르겠습니다... ㅋㅋㅋ

zangsu replied:
우선, 현재의 상황에서 최대한 가독성을 제공하기 위해 트랜잭션 단위의 로직들을 하나의 메서드로 구분해 보았습니다. 
효과가 크진 않아 보이네요... 😅

choijy1705 commented on src/main/java/chess/repository/GameDao.java:
```diff
@@ -0,0 +1,81 @@
+        preparedStatement.execute();
+    }
+
+    public int findLastGameId() {
+        try (Connection connection = connectionGenerator.getConnection()) {
+            PreparedStatement preparedStatement = connection.prepareStatement("SELECT MAX(id) FROM GAMES");
+            ResultSet resultSet = preparedStatement.executeQuery();
+            resultSet.next();
+
+            return resultSet.getInt(1);
+        } catch (final SQLException e) {
+            throw new RuntimeException(e);
+        }
+    }
```
dao에서도 connection 관련로직이 있고
service에서도 있는데
위치를 하나로 통일하는게 더 좋을것 같습니다!

zangsu replied:
위치를 통일시키기 위해 `Service` 에서만 `Connection`을 생성하고, `Dao` 에서는 `Connection`을 파라미터로 넘겨받도록 변경해 주었습니다.

choijy1705 reviewed (changes requested):
짱수 여러가지 시도를 해보고 계시는것 같아요!
미션 막바지인 만큼 다른 시도는 줄이고
도메인과 db table의 차이를 우선적으로 줄여보면 좋을것 같아요!

zangsu committed:
refactor(create_table.sql) : 모든 테이블에 id 값 추가

zangsu committed:
refactor : domain 패키지를 Service 패키지와 분리

zangsu committed:
refactor : SQLException 을 감싸는 언체크 예외를 DatabaseException 으로 변경

zangsu committed:
refactor(TransactionManager) : 조회 쿼리도 Connection 을 받아오도록 변경

zangsu committed:
feat(PieceDao) : Piece 테이블을 분리함에 따라 DAO 추가 생성

zangsu committed:
style : 컨벤션 적용

zangsu committed:
refactor : 가독성을 위해 트랜잭션 단위의 로직들을 메서드로 추출

zangsu committed:
docs : 사용하지 않는 명령어 삭제

zangsu commented on docker/db/mysql/init/create_table.sql:
```diff
@@ -0,0 +1,24 @@
+CREATE TABLE PIECES
```
> board에는 loaction, piece_id 를 가지고 있고 Piece 테이블을 별도로 만들어서 관리하면 도메인을 변경할 필요는 없지 않을까 싶습니다!

영이가 DM으로 해 주셨던 피드백인데요. 무슨 장점을 얻을 수 있는지 이해가 안되어서 추가 학습을 하느라 반영이 조금 오래 걸렸습니다!
아래는 제가 찾아낸, 지금 구현의 장점입니다.

### 1. 수정 이상을 예방할 수 있다.

이전의 구현에선 `piece_type`과 `color` 데이터가 `BOARDS` 테이블에 중복 저장이 되어 있었습니다.
이 때 만약 `piece_type`의 'KING' 이 '킹' 으로 변경된다고 하면 무수히 많은 데이터들을 탐색하며 업데이트를 해 주어야 합니다. 
이 경우 업데이트에서 일어나는 비용이 크기도 하지만, 특정 이유로 단 하나의 데이터라도 업데이트가 되지 않는다면 DB에 잘못된 값이 저장되는 상황으로 이어지게 됩니다. 

지금의 상황에선 단 하나만의 데이터를 수정하면 되기에 수정 이상을 예방하는 동시에, 중복을 제거할 수 있다는 장점이 있겠네요. 
(사실 이 장점은 "오랜 기간 변경되지 않은" 체스 게임이라는 도메인에선 빛을 보지 못하겠지만, 현업에서는 충분히 고려해 볼 수 있을 것 같습니다.)

### 2. 유효성 검사

이제는 새로운 데이터를 입력하는 상황에서도 문자열이 아닌, `PIECES` 테이블의 외래키를 사용하게 됩니다. 
때문에 데이터 입력 과정에서 `PIECES` 테이블에 존재하는 값이 아니라면 오류를 반환하도록 할 수 있다.

(이 경우 외래키 제약 조건을 설정해 주어야 하는 것 같은데요. 추가 학습이 필요할 것 같습니다.)

### 3. 중요한 데이터의 변경 예방

`PIECES` 에 저장될 값은 14개의 가능한 케이스인데요. 이 값을 미리 저장해 두고 일반적인 상황에서 `PIECES` 테이블에 읽기 권한만을 준다면 잘못된 값이 입력되거나, 수정될 상황을 더 확실하게 예방할 수 있을 것 같습니다.


### 마치며

DB를 다루어 본 적이 많이 없어 학습이 제대로 되었는지도 잘 모르겠네요 😅 
혹시 이 외에도 다른 이유가 있다면 학습 키워드 부탁드립니다...ㅎㅎ

choijy1705 replied:
딱히 특별한 이유보다는 도메인과 DB 스키마 간의 불일치를 줄여보기위해 요청드렸습니다🙏
도메인을 변경하려고 하셔서요! 
위와 같은 부분들도 충분히 장점이 있을것 같네요👍👍

zangsu committed:
refactor : 편의성을 위해 dao 패키지 생성

zangsu committed:
refactor(GameDao) : 의미 없는 엔티티 삭제

zangsu committed:
style : 출력문에서 게임 저장 명령어 삭제

zangsu committed:
style : 컨벤션 적용

zangsu reviewed (commented):
지금까지 DB를 많이 다루어 본 적이 없어서 리뷰 반영이 조금 오래 걸린 것 같습니다. 
덕분에 정말 많은 고민들을 새로 하고 배울 수 있었어요!!
이번에도 최선을 다해서 고민을 해 보았습니다. 잘 부탁 드립니다 🙇

choijy1705 commented on src/main/java/chess/service/GameService.java:
```diff
@@ -0,0 +1,135 @@
+        this.transactionManager = transactionManager;
+    }
+
+    public Optional<ChessGame> loadGame() {
+        int lastGameId = transactionManager.getData(gameDao::findLastGameId);
+        Optional<Board> board = transactionManager.getData(
+                connection -> createBoard(connection, lastGameId)
+        );
+        if (board.isEmpty()) {
+            return Optional.empty();
+        }
+
+        Optional<Color> savedTurn = transactionManager.getData(
+                connection -> gameDao.findGameById(connection, lastGameId)
+        );
+        if (savedTurn.isEmpty()) {
+            throw new IllegalStateException("DB에 저장된 턴이 없습니다.");
+        }
+
+        return Optional.of(new PlayingGame(lastGameId, board.get(), savedTurn.get()));
+    }
```
하나의 동작에서 트랜잭션이 이렇게 많다면 
원자성이 보장안되는거 아닐까요?

choijy1705 commented on src/main/java/chess/service/GameService.java:
```diff
@@ -0,0 +1,135 @@
+        }
+    }
+
+    public ChessGame move(ChessGame chessGame, Location source, Location target) {
+        final var movedChessGame = chessGame.move(source, target);
+        transactionManager.executeTransaction(
+                connection -> updateMovementData(connection, source, target, movedChessGame));
+        if (movedChessGame.isEnd()) {
+            transactionManager.executeTransaction(
+                    connection -> deleteGame(connection, movedChessGame.getGameId()));
+        }
+        return movedChessGame;
```
트랜잭션 때문에...너무나도 가독성이 안좋아진것 같아요...😭😭

choijy1705 commented on src/test/java/chess/domain/board/BoardTest.java:
```diff
@@ -0,0 +1,121 @@
+package chess.domain.board;
```
DB관련 test는 없는건가요??

choijy1705 commented on docker/db/mysql/init/create_table.sql:
```diff
@@ -0,0 +1,24 @@
+(
+    id BIGINT NOT NULL AUTO_INCREMENT,
+    game_id BIGINT NOT NULL,
+    location  CHAR(2) NOT NULL,
```
```suggestion
    location CHAR(2) NOT NULL,
```

choijy1705 commented on src/main/resources/database.properties:
```diff
@@ -0,0 +1,9 @@
+#MySQL SERVER ADDRESS
```
test db를 만들거나 혹은 inmemory db를 사용하여 dao관련 테스트도 진행되면 좋을것 같습니다!

choijy1705 commented on src/main/java/chess/service/GameService.java:
```diff
@@ -0,0 +1,135 @@
+                connection -> deleteGame(connection, lastGameId)
+        );
+
+        int newGameId = lastGameId;
+        if (newGameId == 0) {
+            newGameId = 1;
+        }
+        return new InitialGame(newGameId);
```
```suggestion
        
        if (lastGameId == 0) {
            return new InitialGame(1);
        }
        return new InitialGame(lastGameId);
```
어떤 로직인지는 의도가 잘안보이는 로직이네요🤔

zangsu replied:
헉, 의도가 훨씬 잘 드러나네요!!
제안 주신 코드로 변경 후 조건문은 메서드로, '1'은 상수로 추출하여 의도를 조금 더 잘 파악할 수 있도록 변경하겠습니다!

choijy1705 reviewed (approved):
짱수 레벨 1 마지막 미션 체스까지 고생하셨어요!
학습에 대한 의지가 보이는 모습들을 확실히 보여주신것 같아요👍
이제는 이번 미션은 여기서 머지할께요
추가 코멘트 남겼으니 확인부탁드려요🙏
